src/timely-util/src/order.rs

```diff
@@ -0,0 +1,488 @@
+        }
+    }
+
+    /// Returns the timestamp of this partition interval
+    pub fn timestamp(&self) -> &T {
+        &self.0.inner
+    }
+}
+
+impl<P: Partition, T: Timestamp> Timestamp for Partitioned<P, T> {
+    type Summary = PartitionedSummary<P, T>;
+    fn minimum() -> Self {
+        Self(Timestamp::minimum())
+    }
+}
+
+impl<P: Eq, T: PartialOrder> PartialOrder for Partitioned<P, T>
+where
+    Interval<P>: PartialOrder,
+{
+    fn less_equal(&self, other: &Self) -> bool {
+        self.0.less_equal(&other.0)
+    }
+}
+
+/// Helper type alias to access to access the Summary type of a Timestamp implementing type
+type Summary<T> = <T as Timestamp>::Summary;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct PartitionedSummary<P: Partition, T: Timestamp>(
+    Product<Summary<Interval<P>>, Summary<T>>,
+);
+
+impl<P: Partition, T: Timestamp> Default for PartitionedSummary<P, T> {
+    fn default() -> Self {
+        PartitionedSummary(Default::default())
+    }
+}
+
+impl<P: Partition, T: Timestamp> PartialOrder for PartitionedSummary<P, T> {
+    fn less_equal(&self, other: &Self) -> bool {
+        self.0.less_equal(&other.0)
+    }
+}
+
+impl<P: Partition, T: Timestamp> PathSummary<Partitioned<P, T>> for PartitionedSummary<P, T> {
+    fn results_in(&self, src: &Partitioned<P, T>) -> Option<Partitioned<P, T>> {
+        self.0.results_in(&src.0).map(Partitioned)
+    }
+
+    fn followed_by(&self, other: &Self) -> Option<Self> {
+        PathSummary::<Product<Interval<P>, T>>::followed_by(&self.0, &other.0)
+            .map(PartitionedSummary)
+    }
+}
+
+impl<P: Partition, T: Timestamp> PartitionedSummary<P, T> {
+    /// Construct a new summary for a specific partition
+    pub fn with_partition(partition: P, timestamp: Summary<T>) -> Self {
+        Self(Product::new(Interval::Point(partition), timestamp))
+    }
+
+    /// Construct a new summary for an exclusive partition range
+    pub fn with_range(lower: Option<P>, upper: Option<P>, timestamp: Summary<T>) -> Self {
+        let lower = lower.map(RangeBound::Elem).unwrap_or(RangeBound::Bottom);
+        let upper = upper.map(RangeBound::Elem).unwrap_or(RangeBound::Top);
+        Self(Product::new(Interval::Range(lower, upper), timestamp))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+/// A type that represents either an exclusive range or a particular point. When representing a
+/// range either of the two bounds can be minus or positive infinity.
+pub enum Interval<P> {
+    /// A range of points
+    Range(RangeBound<P>, RangeBound<P>),
+    /// A single point
+    Point(P),
+}
+
+impl<P: Ord + Eq> PartialOrder for Interval<P> {
+    fn less_equal(&self, other: &Self) -> bool {
+        use Interval::*;
+        match (self, other) {
+            (Range(self_lower, self_upper), Range(other_lower, other_upper)) => {
+                self_lower <= other_lower && other_upper <= self_upper
```

Comment:
This is implementing `less_equal` so this is why the inequality is with less than or equals. In general, a range that contains another range is less than that range in the ordering. This is why `Range(Bottom, Top)`, the largest range, is the minimum element of this ordering, because there cannot be any other larger range